_posts/2020-01-16-spring-bean-life-cycle.md

```diff
@@ -0,0 +1,161 @@
+          Spring Web MVC controllers and Spring Data repositories."
+image:
+  auto: 0012-pages
+
+---
+
+## What are Spring Beans?
+ > Spring Beans are Java Objects or Instances which is created and managed by Spring container.
+
+Spring `@ApplicationContext` is responsible for creating a Spring bean and manage the 
+Spring bean life cycle.
+
+Below, We have created a simple Spring bean for `Foo` class using `@Component` annotation.
+
+```java
+@Component
+Class Foo { //variables and methods ...}
+```
+## The Spring Bean Lifecycle
+
+The Bean's life cycle consists of 2 phases `Post-initialization` and `Pre-destruction`. 
+
+Below are the Phase 1, `Post-initialization` steps while creating a Bean.
+1. Bean Creation → using Annotations or XML Bean Configs.
+2. Bean Instantiate → In this step Bean will be loaded in `JVM` memory.
+3. Populating Bean properties →  Container will Create a Bean `id`, `scope`, `default values`.
+4. Implementing `BeanNameAware setBeanName()` → Getting to know the Bean Name created by Spring container.
+5. Implementing `BeanClassLoaderAware setBeanClassLoder()` → Getting to know the class is loaded in Spring container Bean Factory.
+6. Implementing `BeanFactoryAware setBeanFactory()` → Getting to know the application Bean and its dependencies. 
+7. Implementing `ApplicationContextAware setApplicationContextAware()` → Getting to know the total number of beans created in the application and etc…
+8. `@PostConstruct` → will be called the annotated method.
+9. Implementing `InitializingBean afterPropertiesSet()` → method will be called after setting the bean properties.
+10. `@Bean init-method` → called specified method inside an `init-method` attribute.
+11. Bean is Ready to use.
+
+Below is Phase 2, `Pre-destruction` steps while destroying a Bean.
+
+1. `@PreDestroy` → Annotated method will be called.
+2. `DisposiableBean destroy()` → will be called at the closing.
+3. `@Bean destroy-method` → called specified method inside the `destroy-method` attribute.
+4. Bean will be destroyed.
+
+## Why Would I Need to Hook into the Bean Lifecycle?
+  Below are some of the use cases.
+  + To make sure application dependency (ex: `Remote Database` and `External Services` and etc...) modules are up and running on startup while creating bean.
+  + Loading application `Metadata` (ex: US state code values) information while creating a bean.
+  + Assigning `Default values` for bean properties (ex: MIN_VALUE and MAX_VALUE).
+  + Scheduling `CRON` jobs or Starting a `Message Queue Listeners` on Bean Creation.
+  + Cleaning up the data which is stored in memory using bean destruction.
+  + Terminating `Process` or `Thread` as part of the bean destruction.
+  + Closing the `Database` connections and `Files`.
+
+
+## Hooking into the Bean Lifecycle
+  Let us look at the 5 different ways of tapping into the spring bean lifecycle.
+## Using Spring @PostConstruct and @PreDestroy annotations

```

Comment:
Just listing the code examples is a little "passionless". We don't need to write an epic tale, but introducing each code snippet with a sentence and then explaining the code snippet in a sentence below it should be the goal.